source/qnx/ionotify_event_loop.c

```diff
@@ -0,0 +1,1042 @@
+     * additional event types converted to CRT event loop domain (enum aws_io_event_type). */
+    int latest_io_event_types;
+    /* Connection opened on the events channel. Used to send pulses to the main event loop. */
+    int pulse_connection_id;
+    struct sigevent event;
+    void *user_data;
+    struct aws_task subscribe_task;
+    struct aws_task resubscribe_task;
+    struct aws_task cleanup_task;
+    /* ID with a value that can fit into pulse user data field (only _NOTIFY_DATA_MASK bits can be used). */
+    int handle_id;
+    /* False when handle is unsubscribed, but this struct hasn't been cleaned up yet. */
+    bool is_subscribed;
+};
+
+/* SI_NOTIFY is a QNX special sigev code requesting resource managers to return active event type along with the event
+ * itself. */
+static short IO_EVENT_PULSE_SIGEV_CODE = SI_NOTIFY;
+static short CROSS_THREAD_PULSE_SIGEV_CODE = _PULSE_CODE_MINAVAIL;
+static short IO_EVENT_KICKSTART_SIGEV_CODE = _PULSE_CODE_MINAVAIL + 1;
+static short IO_EVENT_UPDATE_ERROR_SIGEV_CODE = _PULSE_CODE_MINAVAIL + 2;
+
+/* Setup edge triggered ionotify with a scheduler. */
+struct aws_event_loop *aws_event_loop_new_default_with_options(
+    struct aws_allocator *alloc,
+    const struct aws_event_loop_options *options) {
+    AWS_PRECONDITION(options);
+    AWS_PRECONDITION(options->clock);
+
+    struct aws_event_loop *event_loop = aws_mem_calloc(alloc, 1, sizeof(struct aws_event_loop));
+
+    AWS_LOGF_INFO(AWS_LS_IO_EVENT_LOOP, "id=%p: Initializing edge-triggered ionotify", (void *)event_loop);
+    if (aws_event_loop_init_base(event_loop, alloc, options->clock)) {
+        goto clean_up_loop;
+    }
+
+    struct ionotify_loop *ionotify_loop = aws_mem_calloc(alloc, 1, sizeof(struct ionotify_loop));
+
+    if (options->thread_options) {
+        ionotify_loop->thread_options = *options->thread_options;
+    } else {
+        ionotify_loop->thread_options = *aws_default_thread_options();
+    }
+
+    /* initialize thread id to NULL, it should be updated when the event loop thread starts. */
+    aws_atomic_init_ptr(&ionotify_loop->running_thread_id, NULL);
+
+    aws_linked_list_init(&ionotify_loop->task_pre_queue);
+    ionotify_loop->task_pre_queue_mutex = (struct aws_mutex)AWS_MUTEX_INIT;
+    aws_atomic_init_ptr(&ionotify_loop->stop_task_ptr, NULL);
+
+    if (aws_thread_init(&ionotify_loop->thread_created_on, alloc)) {
+        goto clean_up_ionotify;
+    }
+
+    /* Setup channel to receive events from resource managers. */
+    ionotify_loop->io_events_channel_id = ChannelCreate(0);
+    int errno_value = errno; /* Always cache errno before potential side-effect */
+    if (ionotify_loop->io_events_channel_id == -1) {
+        AWS_LOGF_ERROR(
+            AWS_LS_IO_EVENT_LOOP,
+            "id=%p: ChannelCreate failed with errno %d (%s)\n",
+            (void *)event_loop,
+            errno_value,
+            strerror(errno_value));
```

Comment:
You mean convert errno to crt error and then print crt error string representations?
Considering that in most cases when I got errors I had to lookup errno in the docs anyway to understand why it might be returned by a particular function, I think it's not worth to implement a conversion function. I removed `strerror` for now and kept `errno`, maybe I'll revisit this later on.